builder/googlecompute/step_start_tunnel.go

```diff
@@ -0,0 +1,225 @@
+
+	s.IAPConf.IAPHashBang = fmt.Sprintf("#!%s\n", s.IAPConf.IAPHashBang)
+	log.Printf("[INFO] (google): Prepending inline gcloud setup script with %s",
+		s.IAPConf.IAPHashBang)
+	_, err = writer.WriteString(s.IAPConf.IAPHashBang)
+	if err != nil {
+		return "", fmt.Errorf("Error preparing inline hashbang: %s", err)
+	}
+
+	// authenticate to gcloud
+	_, err = writer.WriteString(
+		fmt.Sprintf("gcloud auth activate-service-account --key-file='%s'\n",
+			s.AccountFile))
+	if err != nil {
+		return "", fmt.Errorf("Error preparing gcloud shell script: %s", err)
+	}
+	// call command
+	args = append([]string{"gcloud"}, args...)
+	argString := strings.Join(args, " ")
+	if _, err := writer.WriteString(argString + "\n"); err != nil {
+		return "", fmt.Errorf("Error preparing gcloud shell script: %s", err)
+	}
+
+	if err := writer.Flush(); err != nil {
+		return "", fmt.Errorf("Error preparing shell script: %s", err)
+	}
+
+	err = os.Chmod(tf.Name(), 0700)
+	if err != nil {
+		log.Printf("[ERROR] (google): error modifying permissions of temp script file: %s", err.Error())
+	}
+
+	// figure out what extension the file should have, and rename it.
+	tempScriptFileName := tf.Name()
+	if s.IAPConf.IAPExt != "" {
+		err := os.Rename(tempScriptFileName, fmt.Sprintf("%s%s", tempScriptFileName, s.IAPConf.IAPExt))
+		if err != nil {
+			return "", fmt.Errorf("Error setting the correct temp file extension: %s", err)
+		}
+		tempScriptFileName = fmt.Sprintf("%s%s", tempScriptFileName, s.IAPConf.IAPExt)
+	}
+
+	return tempScriptFileName, nil
+}
+
+// Run executes the Packer build step that creates an IAP tunnel.
+func (s *StepStartTunnel) Run(ctx context.Context, state multistep.StateBag) multistep.StepAction {
+	if !s.IAPConf.IAP {
+		log.Printf("Skipping step launch IAP tunnel; \"iap\" is false.")
+		return multistep.ActionContinue
+	}
+
+	// shell out to create the tunnel.
+	ui := state.Get("ui").(packer.Ui)
+	instanceName := state.Get("instance_name").(string)
+	c := state.Get("config").(*Config)
+
+	ui.Say("Step Launch IAP Tunnel...")
+
+	err := s.ConfigureLocalHostPort(ctx)
+	if err != nil {
+		state.Put("error", err)
+		ui.Error(err.Error())
+		return multistep.ActionHalt
+	}
+
+	// Generate list of args to use to call gcloud cli.
+	args := []string{"compute", "start-iap-tunnel", instanceName,
+		strconv.Itoa(s.CommConf.Port()),
+		fmt.Sprintf("--local-host-port=localhost:%d", s.IAPConf.IAPLocalhostPort),
+		"--zone", c.Zone,
+	}
+
+	// This is the port the IAP tunnel listens on, on localhost.
+	// TODO make setting LocalHostPort optional
+	s.CommConf.SSHPort = s.IAPConf.IAPLocalhostPort
+
+	log.Printf("Calling tunnel launch with args %#v", args)
+
+	// Create temp file that contains both gcloud authentication, and gcloud
+	// proxy setup call.
+	tempScriptFileName, err := s.createTempGcloudScript(args)
+	if err != nil {
+		state.Put("error", err)
+		ui.Error(err.Error())
+		return multistep.ActionHalt
+	}
+	defer os.Remove(tempScriptFileName)
```

Comment:
The defer needs to be moved up to ensure the file gets cleaned up if an error is encountered.

Reply:
You can disregard this comment. I was mistaken in the execution.